contracts/Registry.sol

```diff
@@ -1,70 +1,604 @@
+
+        orgKeys[indexToUpdate] = orgToUpdate;
+        orgsByName[orgToUpdate].globalOrgIndex = indexToUpdate;
+        orgKeys.length--;
+
+        // delete contents of organization registration
+        delete orgsByName[orgName];
+    }
+
+    //   ____                  _                __  __                 _
+    //  / ___|  ___ _ ____   ___) ___ ___      |  \/  | __ _ _ __ ___ | |_
+    //  \___ \ / _ \ '__\ \ / / |/ __/ _ \     | |\/| |/ _` | '_ ` _ \| __|
+    //   ___) |  __/ |   \ V /| | (__  __/     | |  | | (_| | | | | | | |_
+    //  |____/ \___|_|    \_/ |_|\___\___|     |_|  |_|\__, |_| |_| |_|\__|
+    //                                                 |___/
+
+    function createServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 servicePath, address agentAddress, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, false);
+
+        ServiceRegistration memory service;
+        orgsByName[orgName].servicesByName[serviceName] = service;
+        orgsByName[orgName].servicesByName[serviceName].serviceName = serviceName;
+        orgsByName[orgName].servicesByName[serviceName].servicePath = servicePath;
+        orgsByName[orgName].servicesByName[serviceName].agentAddress = agentAddress;
+        orgsByName[orgName].servicesByName[serviceName].orgServiceIndex = orgsByName[orgName].serviceKeys.length;
+        orgsByName[orgName].serviceKeys.push(serviceName);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function updateServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 servicePath, address agentAddress) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        // update the servicePath and agentAddress
+        orgsByName[orgName].servicesByName[serviceName].servicePath = servicePath;
+        orgsByName[orgName].servicesByName[serviceName].agentAddress = agentAddress;
+    }
+
+    function addTagsToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function addTagToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 tagName) internal {
+
+        // no-op if tag already exists
+        if (orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName == bytes32(0x0)) {
+            // add the service to the org level tag index
```

Comment:
Wait, isn't this `add the tag to the service level tag index`?

Reply:
thats definitely a better description than Lon's. Will change.